x/staking/keeper/val_state_change.go

```diff
@@ -100,15 +106,26 @@ func (k Keeper) ApplyAndReturnValidatorSetUpdates(ctx sdk.Context) (updates []ab
 		validator := k.mustGetValidator(ctx, sdk.ValAddress(valAddrBytes))
 
 		// bonded to unbonding
-		k.bondedToUnbonding(ctx, validator)
+		validator = k.bondedToUnbonding(ctx, validator)
+		amtToNotBonded = amtToNotBonded.Add(validator.GetTokens())
 
 		// delete from the bonded validator index
-		k.DeleteLastValidatorPower(ctx, sdk.ValAddress(valAddrBytes))
+		k.DeleteLastValidatorPower(ctx, validator.GetOperator())
 
 		// update the validator set
 		updates = append(updates, validator.ABCIValidatorUpdateZero())
 	}
 
+	// update the pools based on the recent updates in the validator set
```

Comment:
This block is not trivial at all to understand (in terms of why). Can we please add a more descriptive and elaborate godoc here as to why we need this? Then in each internal case add another godoc. This is arguably one of the most import pieces of code in the state-machine.

Reply:
updated comment